plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServiceViewModel.java

```diff
@@ -85,7 +97,44 @@ private void updateServiceItems(List<ObservableTreeItem> newServiceItems) {
 	}
 
 	protected void updateService(IService service, List<ObservableTreeItem> serviceItems) {
-		firePropertyChange(PROPERTY_SERVICE, null, this.service = getServiceOrDefault(service, serviceItems));
+		IService newService = getServiceOrDefault(service, serviceItems);
+		if(newService != this.service) {
+			boolean needUpdateRoutes = newService == null || this.service == null || newService.getProject() != this.service.getProject();
+			firePropertyChange(PROPERTY_SERVICE, null, this.service = newService);
+			if(needUpdateRoutes) {
+				updateRoutes();
+			}
+		}
+	}
+
+	protected void updateRoutes() {
+		if(this.service == null && !routes.isEmpty()) {
+			List<IRoute> oldRoutes = new ArrayList<>(this.routes);
+			this.routes.clear();
+			firePropertyChange(PROPERTY_ROUTES, oldRoutes, this.routes);
```

Comment:
I removed firing of route change, replaced call to updateRoute() with setRoute(routeToReset) in the end of updateRoutes() and added comments explaining in details why it is necessary.